### PR TITLE
[CARBONDATA-4071] Fix wrong values of date or timestamp child columns on reading through SDK.

### PR DESCRIPTION
**Pull request by Karan980**
### Why is this PR needed?
When date or timestamp column is present inside complex columns (for eg : Array(Date)), it gives wrong result on reading through SDK.
 
 ### What changes were proposed in this PR?
Fix the conversion of INT into date and LONG into timestamp column
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/3339/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/5100/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3347/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5109/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5122/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3360/

File: sdk/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java [hunk 1]
```diff
@@ -126,6 +127,47 @@ public T readNextRow() throws IOException, InterruptedException {
     return formatDateAndTimeStamp((Object []) row);
   }
 
+  public Object getFormattedData(CarbonDimension dimension, Object row, SimpleDateFormat dateFormat,
```
Comment (Indhumathi27) [inline sdk/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java:130]: Add method description

File: sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java [hunk 2]
```diff
@@ -1741,6 +1742,246 @@ public boolean accept(File dir, String name) {
...
+      Assert.assertEquals(i, 10);
+      reader.close();
+      FileUtils.deleteDirectory(new File(path));
+    } catch (Throwable e) {
+      e.printStackTrace();
+      Assert.fail(e.getMessage());
+    }
+  }
+
+  @Test public void testReadDateAndTimestampColumnInStruct()
```
Comment (Indhumathi27) [inline sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java:1833]: can you add a testcase with Map type having Date/Timestamp as key/value

File: sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java [hunk 3]
```diff
@@ -1741,6 +1742,246 @@ public boolean accept(File dir, String name) {
...
+      throws IOException, InvalidLoadOptionException, InterruptedException {
```
Comment (Indhumathi27) [inline sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java:1834]: Remove Exception if not used

File: sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java [hunk 4]
```diff
@@ -1741,6 +1742,246 @@ public boolean accept(File dir, String name) {
...
+      writer.close();
+    } catch (Exception e) {
+      e.printStackTrace();
+      Assert.fail();
+    }
+    Schema schema = CarbonSchemaReader
+        .readSchema(path)
+        .asOriginOrder();
+    assert (4 == schema.getFieldsLength());
+    Field[] fields1 = schema.getFields();
```
Comment (Indhumathi27) [inline sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java:1949]: Remove variable if not used

File: sdk/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java (hunk 1 above)
Comment (Karan980) [reply sdk/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java:141]: Done

File: sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java (hunk 3 above)
Comment (Karan980) [reply sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java]: Done

File: sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java (hunk 4 above)
Comment (Karan980) [reply sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java]: Done

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5162/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3400/

File: sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java (hunk 2 above)
Comment (Karan980) [reply sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java]: Done

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5181/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3419/

File: sdk/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java [hunk 5]
```diff
@@ -126,6 +127,81 @@ public T readNextRow() throws IOException, InterruptedException {
     return formatDateAndTimeStamp((Object []) row);
   }
 
+  /**
+   * This method converts the date and timestamp columns into right format. Before conversion date
+   * is present as integer and timestamp is present as long. This method also flattens complex
+   * columns and format the date/timestamp child present in them.
+   *
+   * @param dimension
```
Comment (Indhumathi27) [inline sdk/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java:135]: Please define the params or else remove it

Comment (Karan980) [reply sdk/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonReader.java]: Done

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5186/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3424/

**Comment by Indhumathi27** [comment]
LGTM

**Comment by VenuReddy2103** [comment]
LGTM

**Comment by Indhumathi27** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5203/

**Comment by Karan980** [comment]
retest this please

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5205/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3445/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/3295/

**Pull request closed without merging**